File: 1081-smallest-subsequence-of-distinct-characters/1081-smallest-subsequence-of-distinct-characters.java
class Solution {
    public String smallestSubsequence(String s) {
        int fre[]=new int[26];
        for(char c:s.toCharArray()){
            fre[c-'a']++;
        }
        boolean visit[]=new boolean[26];
        StringBuilder sb=new StringBuilder();
        for(char c:s.toCharArray()){
            fre[c-'a']--;
            if(visit[c-'a']){
                continue;
            }
            while(sb.length()>0){
                char top=sb.charAt(sb.length()-1);
                if(c<top && fre[top-'a']>0){
                    sb.deleteCharAt(sb.length()-1);
                    visit[top-'a']=false;
                }
                else{
                    break;
                }
            }
            sb.append(c);
            // System.out.println(sb.toString());
            visit[c-'a']=true;
        }
        return sb.toString();
    }
}

File: 1081-smallest-subsequence-of-distinct-characters/1081-smallest-subsequence-of-distinct-characters.py
class Solution:
    def smallestSubsequence(self, s):
        fre = [0] * 26
        for c in s:
            fre[ord(c)-ord('a')]+=1

        visit=[False]*26
        sb=[]

        for c in s:
            fre[ord(c)-ord('a')]-= 1
            if visit[ord(c)-ord('a')]:
                continue
            while sb:
                top=sb[-1]

                if c < top and fre[ord(top) - ord('a')] > 0:
                    sb.pop()
                    visit[ord(top) - ord('a')]=False
                else:
                    break
            sb.append(c)
            visit[ord(c)-ord('a')] = True

        return ''.join(sb)